// nest.js/core/middlewares/routes-mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const router_explorer_1 = require("../router/router-explorer");
const unknown_request_mapping_exception_1 = require("../../errors/exceptions/unknown-request-mapping.exception");
const request_method_enum_1 = require("../../common/enums/request-method.enum");
const shared_utils_1 = require("../../common/utils/shared.utils");
const constants_1 = require("../../common/constants");
const metadata_scanner_1 = require("../metadata-scanner");
class RoutesMapper {
    constructor() {
        this.routerExplorer = new router_explorer_1.ExpressRouterExplorer(new metadata_scanner_1.MetadataScanner());
    }
    mapRouteToRouteProps(routeMetatype) {
        const routePath = Reflect.getMetadata(constants_1.PATH_METADATA, routeMetatype);
        if (shared_utils_1.isUndefined(routePath)) {
            return [this.mapObjectToRouteProps(routeMetatype)];
        }
        const paths = this.routerExplorer.scanForPaths(Object.create(routeMetatype), routeMetatype.prototype);
        return paths.map((route) => ({
            path: this.validateRoutePath(routePath) + this.validateRoutePath(route.path),
            method: route.requestMethod,
        }));
    }
    mapObjectToRouteProps(route) {
        const { path, method } = route;
        if (shared_utils_1.isUndefined(path)) {
            throw new unknown_request_mapping_exception_1.UnknownRequestMappingException();
        }
        return {
            path: this.validateRoutePath(path),
            method: shared_utils_1.isUndefined(method) ? request_method_enum_1.RequestMethod.ALL : method,
        };
    }
    validateRoutePath(path) {
        return shared_utils_1.validatePath(path);
    }
}
exports.RoutesMapper = RoutesMapper;
