// nest.js/core/injector/container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const module_1 = require("./module");
const unkown_module_exception_1 = require("../../errors/exceptions/unkown-module.exception");
class NestContainer {
    constructor() {
        this.modules = new Map();
    }
    addModule(metatype) {
        if (this.modules.has(metatype.name))
            return;
        this.modules.set(metatype.name, new module_1.Module(metatype));
    }
    getModules() {
        return this.modules;
    }
    addRelatedModule(relatedModule, target) {
        if (!this.modules.has(target.name))
            return;
        const module = this.modules.get(target.name);
        const related = this.modules.get(relatedModule.name);
        module.addRelatedModule(related);
    }
    addComponent(component, metatype) {
        if (!this.modules.has(metatype.name)) {
            throw new unkown_module_exception_1.UnkownModuleException();
        }
        const module = this.modules.get(metatype.name);
        module.addComponent(component);
    }
    addExportedComponent(exportedComponent, metatype) {
        if (!this.modules.has(metatype.name)) {
            throw new unkown_module_exception_1.UnkownModuleException();
        }
        const module = this.modules.get(metatype.name);
        module.addExportedComponent(exportedComponent);
    }
    addController(controller, metatype) {
        if (!this.modules.has(metatype.name)) {
            throw new unkown_module_exception_1.UnkownModuleException();
        }
        const module = this.modules.get(metatype.name);
        module.addRoute(controller);
    }
    clear() {
        this.modules.clear();
    }
}
exports.NestContainer = NestContainer;
