// nest.js/core/injector/instance-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const injector_1 = require("./injector");
const logger_service_1 = require("../../common/services/logger.service");
const messages_1 = require("../helpers/messages");
const shared_utils_1 = require("../../common/utils/shared.utils");
class InstanceLoader {
    constructor(container) {
        this.container = container;
        this.injector = new injector_1.Injector();
        this.logger = new logger_service_1.Logger(InstanceLoader.name);
    }
    createInstancesOfDependencies() {
        const modules = this.container.getModules();
        this.createPrototypes(modules);
        this.createInstances(modules);
    }
    createPrototypes(modules) {
        modules.forEach((module) => {
            this.createPrototypesOfComponents(module);
            this.createPrototypesOfRoutes(module);
        });
    }
    createInstances(modules) {
        modules.forEach((module, name) => {
            this.createInstancesOfComponents(module);
            this.createInstancesOfRoutes(module);
            this.callModuleInitHook(module);
            this.logger.log(messages_1.ModuleInitMessage(name));
        });
    }
    createPrototypesOfComponents(module) {
        module.components.forEach((wrapper) => {
            this.injector.loadPrototypeOfInstance(wrapper, module.components);
        });
    }
    createInstancesOfComponents(module) {
        module.components.forEach((wrapper) => {
            this.injector.loadInstanceOfComponent(wrapper, module);
        });
    }
    createPrototypesOfRoutes(module) {
        module.routes.forEach((wrapper) => {
            this.injector.loadPrototypeOfInstance(wrapper, module.routes);
        });
    }
    createInstancesOfRoutes(module) {
        module.routes.forEach((wrapper) => {
            this.injector.loadInstanceOfRoute(wrapper, module);
        });
    }
    callModuleInitHook(module) {
        const components = [...module.routes, ...module.components];
        components.map(([key, { instance }]) => instance)
            .filter((instance) => !shared_utils_1.isNil(instance))
            .filter(this.hasOnModuleInitHook)
            .forEach((instance) => instance.onModuleInit());
    }
    hasOnModuleInitHook(instance) {
        return !shared_utils_1.isUndefined(instance.onModuleInit);
    }
}
exports.InstanceLoader = InstanceLoader;
