// nest.js/websockets/middlewares-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const constants_1 = require("./constants");
const unkown_module_exception_1 = require("../errors/exceptions/unkown-module.exception");
const runtime_exception_1 = require("../errors/exceptions/runtime.exception");
const shared_utils_1 = require("../common/utils/shared.utils");
class MiddlewaresInjector {
    constructor(container) {
        this.container = container;
    }
    inject(server, instance, module) {
        const opaqueTokens = this.reflectMiddlewaresTokens(instance);
        const modules = this.container.getModules();
        if (!modules.has(module)) {
            throw new unkown_module_exception_1.UnkownModuleException();
        }
        const { components } = modules.get(module);
        this.applyMiddlewares(server, components, opaqueTokens);
    }
    reflectMiddlewaresTokens(instance) {
        const prototype = Object.getPrototypeOf(instance);
        return Reflect.getMetadata(constants_1.GATEWAY_MIDDLEWARES, prototype.constructor) || [];
    }
    applyMiddlewares(server, components, tokens) {
        tokens.map(token => this.bindMiddleware(token.name, components))
            .filter(middleware => !shared_utils_1.isNil(middleware))
            .map(middleware => server.use(middleware));
    }
    bindMiddleware(token, components) {
        if (!components.has(token)) {
            throw new runtime_exception_1.RuntimeException();
        }
        const { instance } = components.get(token);
        if (!this.isGatewayMiddleware(instance))
            return null;
        const middleware = instance.resolve();
        return shared_utils_1.isFunction(middleware) ? middleware.bind(instance) : null;
    }
    isGatewayMiddleware(middleware) {
        return !shared_utils_1.isUndefined(middleware.resolve);
    }
}
exports.MiddlewaresInjector = MiddlewaresInjector;
