// nest.js/websockets/socket-server-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const observable_socket_1 = require("./observable-socket");
const io_adapter_1 = require("./adapters/io-adapter");
const shared_utils_1 = require("../common/utils/shared.utils");
class SocketServerProvider {
    constructor(socketsContainer) {
        this.socketsContainer = socketsContainer;
    }
    scanForSocketServer(namespace, port) {
        const observableServer = this.socketsContainer.getServer(namespace, port);
        return observableServer ? observableServer : this.createSocketServer(namespace, port);
    }
    createSocketServer(namespace, port) {
        const server = this.getServerOfNamespace(namespace, port);
        const observableSocket = observable_socket_1.ObservableSocket.create(server);
        this.socketsContainer.addServer(namespace, port, observableSocket);
        return observableSocket;
    }
    getServerOfNamespace(namespace, port) {
        if (namespace) {
            return io_adapter_1.IoAdapter.createWithNamespace(port, this.validateNamespace(namespace));
        }
        return io_adapter_1.IoAdapter.create(port);
    }
    validateNamespace(namespace) {
        return shared_utils_1.validatePath(namespace);
    }
}
exports.SocketServerProvider = SocketServerProvider;
