// nest.js/nest-application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const middlewares_module_1 = require("./core/middlewares/middlewares-module");
const socket_module_1 = require("./websockets/socket-module");
const express_adapter_1 = require("./core/adapters/express-adapter");
const routes_resolver_1 = require("./core/router/routes-resolver");
const logger_service_1 = require("./common/services/logger.service");
const constants_1 = require("./core/constants");
const microservices_module_1 = require("./microservices/microservices-module");
class NestApplication {
    constructor(container, express) {
        this.container = container;
        this.express = express;
        this.logger = new logger_service_1.Logger(NestApplication.name);
        this.routesResolver = new routes_resolver_1.RoutesResolver(container, express_adapter_1.ExpressAdapter);
    }
    setupModules() {
        socket_module_1.SocketModule.setup(this.container);
        middlewares_module_1.MiddlewaresModule.setup(this.container);
        microservices_module_1.MicroservicesModule.setupClients(this.container);
    }
    listen(port, callback) {
        this.setupMiddlewares(this.express);
        this.setupRoutes(this.express);
        this.logger.log(constants_1.messages.APPLICATION_READY);
        return this.express.listen(port, callback);
    }
    setupMiddlewares(instance) {
        middlewares_module_1.MiddlewaresModule.setupMiddlewares(instance);
    }
    setupRoutes(instance) {
        this.routesResolver.resolve(instance);
    }
}
exports.NestApplication = NestApplication;
