// nest.js/nest-microservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const microservices_module_1 = require("./microservices/microservices-module");
const constants_1 = require("./core/constants");
const logger_service_1 = require("./common/services/logger.service");
const server_factory_1 = require("./microservices/server/server-factory");
const transport_enum_1 = require("./common/enums/transport.enum");
class NestMicroservice {
    constructor(container, config) {
        this.container = container;
        this.logger = new logger_service_1.Logger(NestMicroservice.name);
        this.config = Object.assign({ transport: transport_enum_1.Transport.TCP }, config);
        this.server = server_factory_1.ServerFactory.create(this.config);
    }
    setupModules() {
        microservices_module_1.MicroservicesModule.setupClients(this.container);
        microservices_module_1.MicroservicesModule.setupListeners(this.container, this.server);
    }
    listen(callback) {
        this.logger.log(constants_1.messages.APPLICATION_READY);
        this.server.listen(callback);
    }
}
exports.NestMicroservice = NestMicroservice;
