// nest.js/microservices/test/client/client-redis.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sinon = require("sinon");
const chai_1 = require("chai");
const client_redis_1 = require("../../client/client-redis");
describe('ClientRedis', () => {
    const test = 'test';
    const client = new client_redis_1.ClientRedis({});
    describe('getAckPatternName', () => {
        it(`should append _ack to string`, () => {
            const expectedResult = test + '_ack';
            chai_1.expect(client.getAckPatternName(test)).to.equal((expectedResult));
        });
    });
    describe('getResPatternName', () => {
        it(`should append _res to string`, () => {
            const expectedResult = test + '_res';
            chai_1.expect(client.getResPatternName(test)).to.equal((expectedResult));
        });
    });
    describe('sendSingleMessage', () => {
        const pattern = 'test';
        const msg = { pattern };
        let subscribeSpy, publishSpy, onSpy, removeListenerSpy, unsubscribeSpy, sub, pub;
        beforeEach(() => {
            subscribeSpy = sinon.spy();
            publishSpy = sinon.spy();
            onSpy = sinon.spy();
            removeListenerSpy = sinon.spy();
            unsubscribeSpy = sinon.spy();
            sub = {
                subscribe: subscribeSpy,
                on: onSpy,
                removeListener: removeListenerSpy,
                unsubscribe: unsubscribeSpy,
            };
            pub = { publish: publishSpy };
            client.sub = sub;
            client.pub = pub;
        });
        it('should subscribe to response pattern name', () => {
            client.sendSingleMessage(msg, () => { });
            chai_1.expect(subscribeSpy.calledWith(`"${pattern}"_res`)).to.be.true;
        });
        it('should publish stringified message to acknowledge pattern name', () => {
            client.sendSingleMessage(msg, () => { });
            chai_1.expect(publishSpy.calledWith(`"${pattern}"_ack`, JSON.stringify(msg))).to.be.true;
        });
        it('should listen on messages', () => {
            client.sendSingleMessage(msg, () => { });
            chai_1.expect(onSpy.called).to.be.true;
        });
        describe('subscription', () => {
            const callback = sinon.spy();
            const resMsg = {
                err: 'err',
                response: 'test',
            };
            let subscription;
            beforeEach(() => {
                subscription = client.sendSingleMessage(msg, callback);
                subscription(null, JSON.stringify(resMsg));
            });
            it('should call callback with expected arguments', () => {
                chai_1.expect(callback.calledWith(resMsg.err, resMsg.response)).to.be.true;
            });
            it('should unsubscribe to response pattern name', () => {
                chai_1.expect(unsubscribeSpy.calledWith(`"${pattern}"_res`)).to.be.true;
            });
            it('should remove listener', () => {
                chai_1.expect(removeListenerSpy.called).to.be.true;
            });
        });
    });
});
