// nest.js/common/utils/decorators/route-params.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const constants_1 = require("../../constants");
const route_paramtypes_enum_1 = require("../../enums/route-paramtypes.enum");
const assignMetadata = (args, paramtype, index, data) => (Object.assign({}, args, { [`${paramtype}:${index}`]: {
        index,
        data,
    } }));
const createRouteParamDecorator = (paramtype) => {
    return (data) => (target, key, index) => {
        const args = Reflect.getMetadata(constants_1.ROUTE_ARGS_METADATA, target, key) || {};
        Reflect.defineMetadata(constants_1.ROUTE_ARGS_METADATA, assignMetadata(args, paramtype, index, data), target, key);
    };
};
exports.Request = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.REQUEST);
exports.Response = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.RESPONSE);
exports.Next = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.NEXT);
exports.Query = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.QUERY);
exports.Body = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.BODY);
exports.Param = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.PARAM);
exports.Session = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.SESSION);
exports.Headers = createRouteParamDecorator(route_paramtypes_enum_1.RouteParamtypes.HEADERS);
